File: src/main/java/com/financemanager/controllers/LoadManagerController.java
package com.financemanager.controllers;

import com.financemanager.service.AccountManager;
import javafx.event.ActionEvent;
import javafx.fxml.FXML;
import javafx.fxml.FXMLLoader;
import javafx.geometry.Pos;
import javafx.scene.Node;
import javafx.scene.Parent;
import javafx.scene.Scene;
import javafx.scene.control.*;
import javafx.scene.layout.HBox;
import javafx.scene.layout.VBox;
import javafx.stage.FileChooser;
import javafx.stage.Stage;

import java.io.File;
import java.io.IOException;

public class LoadManagerController {

    @FXML private Button back_to_menu_button;
    @FXML private Button editButton;
    @FXML private VBox emptyStateView;
    @FXML private VBox managerListView;

    private boolean isEditMode = false;

    @FXML
    public void initialize() {
        // Listener encadeado: espera que o botão tenha Scene, e a Scene tenha Stage
        // Necessário porque no initialize() o Stage ainda não está disponível
        back_to_menu_button.sceneProperty().addListener((obs, oldScene, newScene) -> {
            if (newScene != null) {
                newScene.windowProperty().addListener((obs2, oldWin, newWin) -> {
                    if (newWin instanceof Stage stage) { // pattern matching - Java 16+
                        stage.setResizable(false);
                        stage.sizeToScene();
                        stage.centerOnScreen();
                    }
                });
            }
        });
        refreshManagerList();
    }

    @FXML
    public void handleToggleEdit(ActionEvent event) {
        isEditMode = !isEditMode;
        editButton.setText(isEditMode ? "Done" : "Edit");
        refreshManagerList();
    }

    private void refreshManagerList() {
        String path = System.getProperty("user.home") + File.separator + "Finance Manager Data" + File.separator + "Managers";
        File folder = new File(path);
        if (!folder.exists()) folder.mkdirs();

        // Lambda como filtro — equivalente ao { _, name -> name.endsWith(".manager") } do Kotlin
        File[] managerFiles = folder.listFiles((dir, name) -> name.endsWith(".manager"));

        if (managerFiles != null && managerFiles.length > 0) {
            toggleViews(true);
            managerListView.getChildren().clear();
            managerListView.setSpacing(15);

            for (File file : managerFiles) {
                String managerName = file.getName().replaceFirst("[.][^.]+$", ""); // nameWithoutExtension

                HBox row = new HBox(10);
                row.setAlignment(Pos.CENTER);

                // --- BOTÃO RENAME (só em Edit Mode) ---
                if (isEditMode) {
                    Button btnRename = new Button("✎");
                    btnRename.setStyle("-fx-background-color: #FFA500; -fx-text-fill: white; -fx-font-weight: bold;");
                    btnRename.setPrefWidth(45);
                    btnRename.setPrefHeight(45);
                    btnRename.setOnAction(e -> handleRename(file));
                    row.getChildren().add(btnRename);
                }

                // --- BOTÃO DO MANAGER ---
                Button btn = new Button(managerName);
                btn.setPrefWidth(isEditMode ? 200 : 300);
                btn.setPrefHeight(45);
                btn.setDisable(isEditMode);
                btn.setOnAction(e -> {
                    AccountManager am = new AccountManager(managerName);
                    AccountManager loaded = am.loadFromFile(file);
                    openDashboard(loaded);
                });
                row.getChildren().add(btn);

                // --- BOTÃO DELETE (só em Edit Mode) ---
                if (isEditMode) {
                    Button btnDelete = new Button("X");
                    btnDelete.setStyle("-fx-background-color: red; -fx-text-fill: white; -fx-font-weight: bold;");
                    btnDelete.setPrefWidth(45);
                    btnDelete.setPrefHeight(45);
                    btnDelete.setOnAction(e -> handleDelete(file));
                    row.getChildren().add(btnDelete);
                }

                managerListView.getChildren().add(row);
            }
        } else {
            toggleViews(false);
            isEditMode = false;
            editButton.setText("Edit");
        }
    }

    private void handleRename(File file) {
        String nameWithoutExt = file.getName().replaceFirst("[.][^.]+$", "");
        TextInputDialog dialog = new TextInputDialog(nameWithoutExt);
        dialog.setTitle("Rename Manager");
        dialog.setHeaderText("Renaming '" + nameWithoutExt + "'");
        dialog.setContentText("New name:");

        dialog.showAndWait().ifPresent(newName -> {
            if (!newName.isBlank()) {
                File newFile = new File(file.getParent(), newName + ".manager");
                if (file.renameTo(newFile)) {
                    AccountManager am = new AccountManager(newName);
                    AccountManager loaded = am.loadFromFile(newFile);
                    loaded.setName(newName);
                    loaded.saveToFile();
                    refreshManagerList();
                }
            }
        });
    }

    private void handleDelete(File file) {
        String nameWithoutExt = file.getName().replaceFirst("[.][^.]+$", "");
        Alert alert = new Alert(Alert.AlertType.CONFIRMATION);
        alert.setTitle("Delete Manager");
        alert.setHeaderText("Confirm deletion of '" + nameWithoutExt + "'");
        alert.setContentText("This will permanently remove the manager file from your computer.");

        if (alert.showAndWait().get() == ButtonType.OK) {
            if (file.delete()) refreshManagerList();
        }
    }

    private void toggleViews(boolean showList) {
        managerListView.setVisible(showList);
        managerListView.setManaged(showList);
        emptyStateView.setVisible(!showList);
        emptyStateView.setManaged(!showList);
    }

    @FXML
    public void handleImport(ActionEvent event) {
        FileChooser fileChooser = new FileChooser();
        fileChooser.setTitle("Import .manager File");
        fileChooser.getExtensionFilters().add(
                new FileChooser.ExtensionFilter("Manager Files", "*.manager")
        );

        File file = fileChooser.showOpenDialog(((Node) event.getSource()).getScene().getWindow());
        if (file != null) {
            String name = file.getName().replaceFirst("[.][^.]+$", "");
            AccountManager am = new AccountManager(name);
            AccountManager imported = am.loadFromFile(file);
            openDashboard(imported);
        }
    }

    @FXML
    public void handleCreate(ActionEvent event) {
        try {
            Parent root = FXMLLoader.load(getClass().getResource("/CreateManager.fxml"));
            navigateTo(root, "Create New Manager", false);
        } catch (IOException e) {
            e.printStackTrace();
        }
    }

    private void openDashboard(AccountManager manager) {
        if (manager == null) return;
        try {
            FXMLLoader loader = new FXMLLoader(getClass().getResource("/Dashboard.fxml"));
            Parent root = loader.load();
            DashboardController controller = loader.getController();
            controller.initData(manager);
            navigateTo(root, "Dashboard", true);
        } catch (IOException e) {
            e.printStackTrace();
        }
    }

    @FXML
    public void handleBacktoMenu(ActionEvent event) {
        try {
            Parent root = FXMLLoader.load(getClass().getResource("/Menu.fxml"));
            navigateTo(root, "Menu", false);
        } catch (IOException e) {
            e.printStackTrace();
        }
    }

    private void navigateTo(Parent root, String title, boolean resizable) {
        Stage stage = (Stage) back_to_menu_button.getScene().getWindow();
        stage.setScene(new Scene(root));
        stage.setTitle(title);
        stage.setResizable(resizable);
        if (!resizable) stage.sizeToScene();
        stage.centerOnScreen();
    }
}

File: src/main/kotlin/com/financemanager/controllers/LoadManagerController.kt
package com.financemanager.controllers

import com.financemanager.service.AccountManager
import javafx.application.Platform
import javafx.event.ActionEvent
import javafx.fxml.FXML
import javafx.fxml.FXMLLoader
import javafx.geometry.Pos
import javafx.scene.Node
import javafx.scene.Parent
import javafx.scene.Scene
import javafx.scene.control.*
import javafx.scene.layout.HBox
import javafx.scene.layout.VBox
import javafx.stage.FileChooser
import javafx.stage.Stage
import java.io.File
import java.io.IOException

class LoadManagerController {

    @FXML lateinit var back_to_menu_button: Button
    @FXML lateinit var editButton: Button
    @FXML lateinit var emptyStateView: VBox
    @FXML lateinit var managerListView: VBox

    private var isEditMode = false

    @FXML
    fun initialize() {
        // Configura a janela (Stage) assim que o controlador ganha acesso à Scene [cite: 2026-01-24]
        back_to_menu_button.sceneProperty().addListener { _, _, newScene ->
            newScene?.windowProperty()?.addListener { _, _, newWin ->
                (newWin as? Stage)?.let { stage ->
                    stage.isResizable = false
                    stage.sizeToScene()
                    stage.centerOnScreen()
                }
            }
        }
        refreshManagerList()
    }

    /**
     * Alterna entre o modo de seleção e o modo de edição (Rename/Delete) [cite: 2026-01-24]
     */
    @FXML
    fun handleToggleEdit(event: ActionEvent) {
        isEditMode = !isEditMode
        editButton.text = if (isEditMode) "Done" else "Edit"
        refreshManagerList()
    }

    /**
     * reconstrói a lista de managers na VBox central
     */
    private fun refreshManagerList() {
        val path = System.getProperty("user.home") + File.separator + "Finance Manager Data" + File.separator + "Managers"
        val folder = File(path)
        if (!folder.exists()) folder.mkdirs()

        val managerFiles = folder.listFiles { _, name -> name.endsWith(".manager") }

        if (managerFiles != null && managerFiles.isNotEmpty()) {
            toggleViews(true)
            managerListView.children.clear()
            managerListView.spacing = 15.0

            managerFiles.forEach { file ->
                val managerName = file.nameWithoutExtension

                // HBox para conter os botões da linha (Simula o rowPanel do Swing)
                val row = HBox(10.0).apply {
                    alignment = Pos.CENTER
                }

                // --- BOTÃO RENAME (Só aparece em Edit Mode) ---
                if (isEditMode) {
                    val btnRename = Button("✎").apply {
                        style = "-fx-background-color: #FFA500; -fx-text-fill: white; -fx-font-weight: bold;"
                        prefWidth = 45.0
                        prefHeight = 45.0
                        setOnAction { handleRename(file) }
                    }
                    row.children.add(btnRename)
                }

                // --- BOTÃO DO MANAGER ---
                val btn = Button(managerName).apply {
                    prefWidth = if (isEditMode) 200.0 else 300.0
                    prefHeight = 45.0
                    isDisable = isEditMode // Desativa cliques para abrir enquanto edita
                    setOnAction {
                        // CHAMADA DE INSTÂNCIA COM OBJETO FILE [cite: 2026-01-24]
                        val am = AccountManager(managerName)
                        val loadedManager = am.loadFromFile(file)
                        openDashboard(loadedManager)
                    }
                }
                row.children.add(btn)

                // --- BOTÃO DELETE (Só aparece em Edit Mode) ---
                if (isEditMode) {
                    val btnDelete = Button("X").apply {
                        style = "-fx-background-color: red; -fx-text-fill: white; -fx-font-weight: bold;"
                        prefWidth = 45.0
                        prefHeight = 45.0
                        setOnAction { handleDelete(file) }
                    }
                    row.children.add(btnDelete)
                }

                managerListView.children.add(row)
            }
        } else {
            toggleViews(false)
            isEditMode = false
            editButton.text = "Edit"
        }
    }

    private fun handleRename(file: File) {
        val dialog = TextInputDialog(file.nameWithoutExtension)
        dialog.title = "Rename Manager"
        dialog.headerText = "Renaming '${file.nameWithoutExtension}'"
        dialog.contentText = "New name:"

        dialog.showAndWait().ifPresent { newName ->
            if (newName.isNotBlank()) {
                val newFile = File(file.parent, "$newName.manager")
                if (file.renameTo(newFile)) {
                    // Instanciamos e carregamos do novo ficheiro para atualizar o nome interno [cite: 2026-01-24]
                    val am = AccountManager(newName)
                    val loaded = am.loadFromFile(newFile)
                    loaded.setName(newName) // Sincroniza o campo name
                    loaded.saveToFile()      // Atualiza o ficheiro com o novo estado
                    refreshManagerList()
                }
            }
        }
    }

    private fun handleDelete(file: File) {
        val alert = Alert(Alert.AlertType.CONFIRMATION).apply {
            title = "Delete Manager"
            headerText = "Confirm deletion of '${file.nameWithoutExtension}'"
            contentText = "This will permanently remove the manager file from your Fedora."
        }

        if (alert.showAndWait().get() == ButtonType.OK) {
            if (file.delete()) refreshManagerList()
        }
    }

    private fun toggleViews(showList: Boolean) {
        managerListView.isVisible = showList
        managerListView.isManaged = showList
        emptyStateView.isVisible = !showList
        emptyStateView.isManaged = !showList
    }

    @FXML
    fun handleImport(event: ActionEvent) {
        val fileChooser = FileChooser().apply {
            title = "Import .manager File"
            extensionFilters.add(FileChooser.ExtensionFilter("Manager Files", "*.manager"))
        }
        val file = fileChooser.showOpenDialog((event.source as Node).scene.window)

        if (file != null) {
            // CHAMADA DE INSTÂNCIA COM OBJETO FILE [cite: 2026-01-24]
            val am = AccountManager(file.nameWithoutExtension)
            val imported = am.loadFromFile(file)
            openDashboard(imported)
        }
    }

    @FXML
    fun handleCreate(event: ActionEvent) {
        val root = FXMLLoader.load<Parent>(javaClass.getResource("/CreateManager.fxml"))
        navigateTo(root, "Create New Manager", resizable = false)
    }

    private fun openDashboard(manager: AccountManager?) {
        if (manager == null) return
        val loader = FXMLLoader(javaClass.getResource("/Dashboard.fxml"))
        val root = loader.load<Parent>()
        val controller = loader.getController<DashboardController>()
        controller.initData(manager)
        navigateTo(root, "Dashboard", resizable = true)
    }

    @FXML
    fun handleBacktoMenu(event: ActionEvent) {
        val root = FXMLLoader.load<Parent>(javaClass.getResource("/Menu.fxml"))
        navigateTo(root, "Menu", resizable = false)
    }

    private fun navigateTo(root: Parent, titleStr: String, resizable: Boolean) {
        val stage = back_to_menu_button.scene.window as Stage
        stage.scene = Scene(root)
        stage.title = titleStr
        stage.isResizable = resizable
        if (!resizable) stage.sizeToScene()
        stage.centerOnScreen()
    }
}